### PR TITLE
Validate naked functions definitions

### PR DESCRIPTION
**Pull request by tmiasko**
Validate that naked functions are defined in terms of a single inline assembly
block that uses only `const` and `sym` operands and has `noreturn` option.

Implemented as future incompatibility lint with intention to migrate it into
hard error. When it becomes a hard error it will ensure that naked functions are
either unsafe or contain an unsafe block around the inline assembly. It will
guarantee that naked functions do not reference functions parameters (obsoleting
part of existing checks from #79411). It will limit the definitions of naked
functions to what can be reliably supported. It will also reject naked functions
implemented using legacy LLVM style assembly since it cannot satisfy those
conditions.

https://github.com/rust-lang/rfcs/pull/2774
https://github.com/rust-lang/rfcs/pull/2972

### CODE REVIEW

File: compiler/rustc_lint_defs/src/builtin.rs [hunk 1]
```diff
@@ -2742,6 +2742,40 @@ declare_lint! {
...
+    /// While other definitions of naked functions were previously accepted, they
+    /// are unsupported and might not work reliably. This is a [future-incompatible]
+    /// lint that will transition into hard error in the future.
+    ///
+    /// [future-incompatible]: ../index.md#future-incompatible-lints
+    pub UNSUPPORTED_NAKED_FUNCTIONS,
+    Warn,
+    "unsupported naked function definitions",
+    @future_incompatible = FutureIncompatibleInfo {
+        reference: "issue #????? <https://github.com/rust-lang/rust/issues/?????>",
```
Comment (Amanieu) [inline compiler/rustc_lint_defs/src/builtin.rs:2774]: Maybe point it to #32408

File: compiler/rustc_passes/src/naked_functions.rs [hunk 2]
```diff
@@ -103,11 +109,173 @@ impl<'tcx> Visitor<'tcx> for CheckBody<'tcx> {
...
+                UNSUPPORTED_NAKED_FUNCTIONS,
+                hir_id,
+                unsupported_ops,
+                |lint| {
+                    lint.build("only `const` and `sym` operands are supported in naked functions")
+                        .emit();
+                },
+            );
+        }
+        if !asm.options.contains(InlineAsmOptions::NORETURN) {
```
Comment (Amanieu) [inline compiler/rustc_passes/src/naked_functions.rs:251]: I would go even further here: any option other than `att_syntax` is invalid in this context.

**Review by Amanieu — COMMENTED** [review]
Do we want to also allow `llvm_asm!`? This would be without any parameter validation.

**Comment by tmiasko** [comment]
> Do we want to also allow `llvm_asm!`? This would be without any parameter validation.

Given that this starts as a future incompatibility lint, with possibility for migration without causing immediate breakage, and all features involved are still unstable, I would lean towards rejecting `llvm_asm` as reflected in the current implementation. Though, I can of course change that, in that case I would also allow unreachable intrinsic.

File: compiler/rustc_passes/src/naked_functions.rs (hunk 2 above)
Comment (Amanieu) [reply compiler/rustc_passes/src/naked_functions.rs:251]: Can you extend this to reject options other than `att_syntax` and `noreturn`?

Comment (tmiasko) [reply compiler/rustc_passes/src/naked_functions.rs:291]: I extended validation to reject options other than `noreturn` and `att_syntax`.

**Comment by tmiasko** [comment]
* Updated option validation.
* Added check that non-Rust ABI is used.

File: compiler/rustc_passes/src/naked_functions.rs [hunk 3]
```diff
@@ -103,11 +134,188 @@ impl<'tcx> Visitor<'tcx> for CheckBody<'tcx> {
...
+                        .emit();
+                },
+            );
+        }
+
+        let unsupported_options: Vec<&'static str> = [
+            (InlineAsmOptions::NOMEM, "`nomem`"),
+            (InlineAsmOptions::PRESERVES_FLAGS, "`preserves_flags`"),
+            (InlineAsmOptions::PURE, "`pure`"),
+            (InlineAsmOptions::READONLY, "`readonly`"),
```
Comment (Amanieu) [inline compiler/rustc_passes/src/naked_functions.rs:275]: I would `no_stack` to this list.

Comment (tmiasko) [reply compiler/rustc_passes/src/naked_functions.rs:276]: That was unintentional omission; fixed. Thanks.

**Comment by Amanieu** [comment]
You need to mark the test with `// only-x86_64`. Otherwise you get errors on target that don't support `asm!`.

**Comment by tmiasko** [comment]
Ignored test on wasm32. The other targets should be fine, I think? (the att_syntax test is conditional, last, and produces no errors).

**Comment by Amanieu** [comment]
Not exactly, we have plenty of targets which don't support `asm!` (e.g. sparc, s390, etc). I would prefer if the test was x86_64-only. It's fine since this feature isn't architecture-specific.

**Comment by tmiasko** [comment]
Sure, replaced with `only-x86_64`.

**Comment by Acciente717** [comment]
Could someone please tell me why naked functions should be `noreturn`?

**Comment by lzutao** [comment]
@Acciente717 That's explained here: https://www.reddit.com/r/rust/comments/kwn43k/why_must_naked_functions_not_return/gj59yqn/

But it would be better to document it officially ?

**Comment by Acciente717** [comment]
Yeah, I think so.

Also, it seems that I misunderstood the `noreturn` requirement. Now my understanding is that
1. We should add `options(noreturn)` to the `asm!` block. This will prevent the compiler from automatically generating the return instruction.
2. The function marked as `#[naked]` does not necessarily need to have the return signature `-> !`, as long as we make sure that the return value is correctly passed back to the caller conforming to the calling convention in the `asm!` block.

Are these correct?

**Comment by valff** [comment]
Is this use-case with type parameters has been banned intentionally or is it a bug?

```rust
#![feature(asm)]
#![feature(naked_functions)]

pub trait Foo {
    const X: usize;
}

#[naked]
pub extern "C" fn foo<T: Foo>() {
    unsafe {
        asm!(
            "mov r8, {}",
            in(reg) T::X,
            options(noreturn),
        );
    }
}

fn main() {
    struct Bar;
    impl Foo for Bar {
        const X: usize = 42;
    }
    foo::<Bar>();
}
```

([Playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=7b93310b7080ac993a8cf5c4d0f25de8))

**Pull request merged**